### ISSUE TITLE
Big variance for the first class "Algebra"

### ISSUE THREAD

**Issue by pavel-collab** [pavel-collab/SBT-kaggle-nlp#8]
In our implementation classifier fall in the first label. It often give some another prediction for this examples. I think, it's because samples with label algebra is quite similar to some other classes. There are some ways how we can solve this problem

1/ we can give learned model, freeze some weights and learn only classification head.
2/ we can change class weights to make the first class more heavy
3/ Chat GPT suggest to use Focal Loss

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class FocalLoss(nn.Module):
    def __init__(self, alpha=None, gamma=2.0, reduction='mean'):
        """
        alpha: Веса для классов (можно поставить приоритет на какой-то класс)
        gamma: Насколько сильно фокусироваться на ошибках (обычно 2.0)
        """
        super(FocalLoss, self).__init__()
        self.alpha = alpha
        self.gamma = gamma
        self.reduction = reduction

    def forward(self, logits, labels):
        ce_loss = F.cross_entropy(logits, labels, weight=self.alpha, reduction='none')
        pt = torch.exp(-ce_loss)  # вероятность предсказания правильного класса
        focal_loss = ((1 - pt) ** self.gamma) * ce_loss

        if self.reduction == 'mean':
            return focal_loss.mean()
        elif self.reduction == 'sum':
            return focal_loss.sum()
        else:
            return focal_loss
```

**Comment by pavel-collab** [comment]
custom model arch
```
class BertWithFocalLoss(BertPreTrainedModel):
    def __init__(self, config, focal_loss=None):
        super().__init__(config)
        self.bert = BertModel(config)
        self.dropout = nn.Dropout(config.hidden_dropout_prob)
        self.classifier = nn.Linear(config.hidden_size, config.num_labels)
        self.focal_loss = focal_loss or FocalLoss()

        self.init_weights()

    def forward(self, input_ids=None, attention_mask=None, labels=None, **kwargs):
        outputs = self.bert(input_ids, attention_mask=attention_mask, **kwargs)
        pooled_output = outputs[1]  # CLS токен
        pooled_output = self.dropout(pooled_output)
        logits = self.classifier(pooled_output)

        loss = None
        if labels is not None:
            loss = self.focal_loss(logits, labels)

        return {"loss": loss, "logits": logits}
```

example of training
```
model = BertWithFocalLoss.from_pretrained("bert-base-uncased", num_labels=4)

training_args = TrainingArguments(
    output_dir="./results",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=32,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    logging_dir="./logs",
    num_train_epochs=3,
    logging_steps=100,
    load_best_model_at_end=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=encoded_dataset["train"],
    eval_dataset=encoded_dataset["test"],
    compute_metrics=compute_metrics,
    tokenizer=tokenizer,
)

trainer.train()
```
